### ISSUE TITLE
FileLink accepts absolute path but doesn't work

### ISSUE THREAD

**Issue by vbraun** [ipython/ipython#8053]
The following shows a link, but clicking on the link yields a 404:

    from IPython.display import FileLink
    FileLink('/tmp/foo.png')

Also, the FileLink docs do not mention any restrictions on which files can be linked. Its either a bug or lacking documentation + input checking.

Ideally you could also set headers for the download link, in particular content-type and maybe cache-control.

**Comment by vbraun** [comment]
Somewhat related, relative paths in FileLink are normalized incorrectly, for example

    FileLink('../../tmp/foo.png')

displays `http://localhost:8888/tmp/foo.png` as link which obviously can't work. 

If I manually browse to a url with a relative path component like `http://localhost:8888/files/../../tmp/foo.png` then I'm getting redirected to the incorrect path normalization  `http://localhost:8888/tmp/foo.png` which again can't work. Either forbid relative paths (why?) or don't normalize it.

**Comment by minrk** [comment]
Yup, we should raise on any attempt to link to absolute files, since absolute paths will always fail due to URL tree isolation (only files within the root server dir are accessible).

**Comment by vbraun** [comment]
I haven't seen that restriction spelled out, nor does it make much sense to me. Why should only files in the tree descending from the ipython cwd be accessible? There is no security benefit. Temporary files are almost always outside of it and are a likely candidate for files that you might want to download from a remote server.

**Comment by minrk** [comment]
@vbraun we don't handle the path, browsers do that. `..` behaves the same in URLs as it does in a path, so `/files/../foo` results in a request for `/foo`, and `/files/path/../otherpath` results in `/files/otherpath`, which is correct. Because of this, relative paths do actually work correctly, as long as they don't leave the root of the served directory.

The kernel, where the links are computed, doesn't know the server's root directory, so it can't verify that the root has been escaped.

**Comment by minrk** [comment]
@vbraun the contents service, which serves these files, is restricted to a particular path. It does not serve all files on the system, nor should it.

**Comment by vbraun** [comment]
Ah good point, the handler doesn't see the '..'

Still, why shouldn't it server all files? You can already execute arbitrary code. The standard tornado file server shouldn't of course, but the standard reason clearly doesn't apply.

**Comment by Carreau** [comment]
Kernels and server might not be on the same machine. You could run a docker instace isolated, then you don't care wether or not it runs arbitrary code. You don't want the server to expose your `~/.ssh/` keys.

**Comment by vbraun** [comment]
Ok, I was under the impression that FileLink would serve from the kernel filesystem in that case. If its serving from the notebook server filesystem then its pretty useless to me anyways.

**Comment by minrk** [comment]
You can make it serve all files:

    ipython notebook --notebook-dir=/

**Comment by minrk** [comment]
Yes, FileLinks came about only as a convenient shortcut for building common URLs. It doesn't serve any files itself, and will be totally useless if the contents service and kernels are separated.

**Comment by vbraun** [comment]
So what is the recommended way for getting at files that the the kernel process created?

**Comment by minrk** [comment]
We don't have a generic mechanism for serving files from the kernel if it is remote from the notebook. You can use the display system to publish things directly to the frontend, but there is no mechanism for URLs to refer to kernel data.

Are you planning on separating notebook servers from their kernels? This is a rare, and generally not recommended pattern, because it comes with a lot of headaches, and few benefits. Isolating the notebook server and kernels together as a single multi-process application is more often the way to go.

**Comment by vbraun** [comment]
I'm trying to find a way for Sage users to download generated files (where most types are not supported by the IPython display system) without having to know how their administrator set up the server.

**Comment by minrk** [comment]
I don't know about the Sage deployment, but Sage could add an additional StaticFileHandler, serving all files in the kernel's filesystem. This should only be done if access to the handler itself is as restricted as access to execution on a kernel.
